Compile/src/Producers/PrefiringWeightProducer.cc

```diff
@@ -1,162 +1,338 @@
+                    continue;
+                double prefiringprob_mu = getPrefiringRateMuon(eta, phi, pt, fluct);
+                nonPrefiringProbaMuon[fluct] *= (1. - prefiringprob_mu);
             }
-            //Last case: if overlapping photons have a non prefiring rate smaller than the jet, don't consider the jet in the event weight, and do nothing.
         }
     }
+    //Move global prefire weights, as well as those for muons, photons, and jets, to the event
+    //Up and Down are switched, since we use the inverse and apply it to data.
     product.m_optionalWeights["prefiringWeight"] =  1./nonPrefiringProba[0];
-    product.m_optionalWeights["prefiringWeightUp"] =  1./nonPrefiringProba[1];
-    product.m_optionalWeights["prefiringWeightDown"] =  1./nonPrefiringProba[2];
-  
+    product.m_optionalWeights["prefiringWeightUp"] =  1./nonPrefiringProba[2];
+    product.m_optionalWeights["prefiringWeightDown"] =  1./nonPrefiringProba[1];
+
+    product.m_optionalWeights["prefiringWeightECAL"] =  1./nonPrefiringProbaECAL[0];
+    product.m_optionalWeights["prefiringWeightECALUp"] =  1./nonPrefiringProbaECAL[2];
+    product.m_optionalWeights["prefiringWeightECALDown"] =  1./nonPrefiringProbaECAL[1];
+
+    product.m_optionalWeights["prefiringWeightMuon"] =  1./nonPrefiringProbaMuon[0];
+    product.m_optionalWeights["prefiringWeightMuonUp"] =  1./nonPrefiringProbaMuon[2];
+    product.m_optionalWeights["prefiringWeightMuonDown"] =  1./nonPrefiringProbaMuon[1];
+    product.m_optionalWeights["prefiringWeightMuonUpStat"] =  1./nonPrefiringProbaMuon[4];
+    product.m_optionalWeights["prefiringWeightMuonDownStat"] =  1./nonPrefiringProbaMuon[3];
+    product.m_optionalWeights["prefiringWeightMuonUpSyst"] =  1./nonPrefiringProbaMuon[6];
+    product.m_optionalWeights["prefiringWeightMuonDownSyst"] =  1./nonPrefiringProbaMuon[5];
 }
 
+double PrefiringWeightProducer::getPrefiringRateEcal(double eta,
+                                                     double pt,
+                                                     TH2F* h_prefmap,
+                                                     fluctuations fluctuation) const {
+    //Check pt is not above map overflow
+    int nbinsy = h_prefmap->GetNbinsY();
+    double maxy = h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy + 1);
+    if (pt >= maxy)
+        pt = maxy - 0.01;
+    int thebin = h_prefmap->FindBin(eta, pt);
+
+    double prefrate = h_prefmap->GetBinContent(thebin);
+
+    double statuncty = h_prefmap->GetBinError(thebin);
+    double systuncty = prefiringRateSystUncEcal_ * prefrate;
+
+    if (fluctuation == up)
+        prefrate = std::min(1., prefrate + sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+    else if (fluctuation == down)
+        prefrate = std::max(0., prefrate - sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+    if (prefrate > 1.) {
+        LOG(WARNING) << this->GetProducerId() << ": " << "Found a prefiring probability > 1. Setting to 1.";
```

Comment:
How can this happen? :D

Reply:
This is from the original code. I did not look into the histograms and did not check the values.

Reply:
So this may happen, but then this workaround is recommended I guess as it's the same in CMSSW.

Reply:
Okay, thanks.